### PR TITLE
Swift: Add new query for XML External Entities (XML) vulnerabilities

### PR DESCRIPTION
**Pull request by atorralba**
Adds a new query to cover XXE vulnerabilities. Only covers `XMLParser` sinks for now, follow up PRs will add more sinks in third-party libraries.

Note that the CSV summaries added for the classes `Data` and `InputStream` are the bare minimum for the XXE tests to work properly. These classes need proper modeling, which deserve their own PRs.

### CODE REVIEW

File: swift/ql/test/query-tests/Security/CWE-611/testXXE.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+// --- tests ---
+
+func testData() {
+    let remoteString = String(contentsOf: URL(string: "http://example.com/")!)
+    let remoteData = Data(remoteString)
+    let parser = XMLParser(data: remoteData) // $ hasXXE=32
+    parser.shouldResolveExternalEntities = true
+}
```
Comment (geoffw0) [inline swift/ql/test/query-tests/Security/CWE-611/testXXE.swift:36]: I notice that we don't care where the parser is used after its set up incorrectly.  I think its fair to assume that it will be in the case above, why else would it be created like that?  But for libraries where the default configuration is unsafe there's a plausible edge case:
```
let parser = UnsafeByDefaultParser(data: remoteData)
parser.parse(...) // BAD
parser.shouldResolveExternalEntities = false // makes it safe, but the above is still BAD
```
How concerned are you about this case?  I believe the CPP version of the query does account for it, I'm not entirely sure if its worth the effort though.

File: swift/ql/lib/codeql/swift/security/XXE.qll [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/** Provides classes and predicates to reason about XML external entities (XXE) vulnerabilities. */
+
+import swift
+private import codeql.swift.dataflow.DataFlow
+private import codeql.swift.dataflow.internal.DataFlowPrivate
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/security/XXE.qll:5]: What is `DataFlowPrivate` need for?

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import swift
+private import codeql.swift.dataflow.ExternalFlow
+
+private class DataSummaries extends SummaryModelCsv {
+  override predicate row(string row) { row = ";Data;true;init(_:);;;Argument[0];ReturnValue;taint" }
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll:5]: > These classes need proper modeling, which deserve their own PRs.

I agree - are you planning to do this?  Shall I create an issue for it?

File: swift/ql/test/query-tests/Security/CWE-611/testXXE.swift [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    let remoteStream = InputStream(data: remoteData)
+    let parser = XMLParser(stream: remoteStream) // $ hasXXE=39
+    parser.shouldResolveExternalEntities = true
+
+}
+
+func testDataSafe() {
+    let remoteString = String(contentsOf: URL(string: "http://example.com/")!)
+    let remoteData = Data(remoteString)
+    let _ = XMLParser(data: remoteData) // NO XXE: parser doesn't enable external entities
```
Comment (geoffw0) [inline swift/ql/test/query-tests/Security/CWE-611/testXXE.swift:50]: Note that there is also an `XMLParser.init?(contentsOf url: URL)` method.

File: swift/ql/test/query-tests/Security/CWE-611/testXXE.swift (hunk 1 above)
Comment (atorralba) [reply swift/ql/test/query-tests/Security/CWE-611/testXXE.swift:36]: I expected to model other libraries differently (since the `shouldResolveExternalEntities` attribute will probably have a different name anyway). But in any case, the current implementation requires `true` to be set to `shouldResolveExternalEntities`, so setting it to `false` shouldn't flag the parser as vulnerable.

To be 100% sure, I added a couple of test cases in e1aa75a.

File: swift/ql/lib/codeql/swift/security/XXE.qll (hunk 2 above)
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: It's for being able to use the `Unit` class `XxeAdditionalTaintStep` extends. Other languages (e.g. Java) expose this class in the public `DataFlow` API, but it's not the case for Swift. Should we fix that instead of using this import?

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll (hunk 3 above)
Comment (atorralba) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll:5]: I can work on that after finishing this query if you think it has enough priority. An issue for it would be nice in any case.

File: swift/ql/test/query-tests/Security/CWE-611/testXXE.swift (hunk 4 above)
Comment (atorralba) [reply swift/ql/test/query-tests/Security/CWE-611/testXXE.swift:56]: That's a good point, and it also made me realize that I needed to handle optionals in `XmlParserRef`. See fix in `f9c49bb`, thanks!

File: swift/ql/test/query-tests/Security/CWE-611/testXXE.swift (hunk 1 above)
Comment (atorralba) [reply swift/ql/test/query-tests/Security/CWE-611/testXXE.swift:36]: > I notice that we don't care where the parser is used after its set up incorrectly.

We could require a [`parse()`](https://developer.apple.com/documentation/foundation/xmlparser/1411778-parse) call with even more local data flow, but my concern is that we'd miss cases where local flow isn't enough to track the vulnerable parser to the location where `parse()` is called. And, as you said, normally a parser wouldn't be initialized to then not be used (and still, if that was the case, it would be a vulnerability waiting to happen which should be fixed IMHO).

File: swift/ql/lib/codeql/swift/security/XXE.qll (hunk 2 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: Its a bit weird to use a class from dataflow for this, as it seems to have nothing to do with dataflow.  Seems like ideally it should be defined somewhere outside.  I guess dataflow itself doesn't want to rely on that.

Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: Java defines a `Unit.qll` file outside of the dataflow folder, and imports it wherever it's needed. Let me do just that for Swift so that we remove this dependency.

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll (hunk 3 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll:5]: I do think its important.  Do you think covering `NSData` as well should have similar or lower priority?

File: swift/ql/lib/codeql/swift/security/XXE.qll (hunk 2 above)
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: https://github.com/github/codeql/pull/11096. I'll rebase when it's merged.

Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: This is probably because Java (like many other of our languages) applies the pattern of, for each query, defining a `.qll` file containing (1) abstract classes for the sources, sinks, and sanitizers, and (2) a mechanism for extending the step relation of the configuration (which is the unit class trick we're doing here).

This pattern makes it possible for users to extend the query's behavior without modifying the `.ql` file at all, which is really nice!

This is not a pattern we've applied yet in Swift (since we never settled on whether we were okay with the foot gun related to having dataflow/taintflow configurations in `.qll` files). Now that we have a ql-for-ql query to warn us when this footgun is fired, I think we should consider following the pattern sketched in this PR more consistently for future queries (and possible rewrite our existing queries to follow this. This could be considered tech debt).

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll (hunk 3 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll:5]: I've created https://github.com/github/codeql-c-team/issues/1334 for modelling `Data`; please update the description as you see fit.

File: swift/ql/lib/codeql/swift/security/XXE.qll (hunk 2 above)
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: Sounds good. But I don't want to force my habits from another language if they don't really fit here. Let me know if you want me to move everything back to the `.ql` file as the other queries do.

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll (hunk 3 above)
Comment (atorralba) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll:5]: > Do you think covering NSData as well should have similar or lower priority?

I think they probably should have similar priority.

File: swift/ql/lib/codeql/swift/security/XXE.qll (hunk 2 above)
Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: I really like the pattern, so I think it's great to have a template query that we can copy/paste from in the future 😄. It would be good to have @geoffw0's opinion on this as well, though.

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: I don't mind the query in a separate `.qll` pattern, it has advantages for testing and extension.  The disadvantage seems to be a proliferation of files making things a bit harder to navigate, especially when some are in the `src` directory and some in `lib`.  It feels like `lib/codeql/swift/security` could become a mess if we do all taint queries exactly like this one???

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll (hunk 3 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll:5]: I'll create an issue for `NSData` (and `NSString`) as well then.

File: swift/ql/lib/codeql/swift/security/XXE.qll (hunk 2 above)
Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/XXE.qll:5]: It would look something like this: https://github.com/github/codeql/tree/main/java/ql/lib/semmle/code/java/security

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/XXE.qll]: I think I would prefer if it had the same directory structure as `src` and `tests`; though I suppose that would mean `CWE-XYZ` files with only one file (`.qll`) in each directory.  So perhaps I haven't thought this through enough.

Comment (atorralba) [reply swift/ql/lib/codeql/swift/security/XXE.qll]: The bad thing about that is that `CWE-XYZ` directories wouldn't be importable AFAIK.

Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/security/XXE.qll]: > The bad thing about that is that CWE-XYZ directories wouldn't be importable AFAIK.

Correct :(

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/XXE.qll]: OK, I'm happy with the way you've arranged this query.  At some point, perhaps after a few weeks of getting used to this way of doing things, we can decide if we want to make the other (dataflow) queries consistent with it (or not).

**Comment by geoffw0** [comment]
I think this might be ready for docs review?

**Review by mchammer01 — DISMISSED** [review]
@atorralba - this LGTM ✨ 
A few minor suggestions 🙂

**Comment by atorralba** [comment]
Hey @mchammer01 thanks for the review! Actually, the QHelp file was copy-pasted from the XXE query of other languages, but still I applied your suggestions. We can always apply those same fixes to the other languages in a follow-up PR 😄

**Comment by mchammer01** [comment]
Oh sorry @atorralba - I didn't know that 🙈

**Review by mchammer01 — APPROVED** [review]
LGTM 💖

**Review by geoffw0 — APPROVED** [review]

**Comment by geoffw0** [comment]
DCA run looks good to me (aside from some noise that has nothing to do with the new query). :+1:
